src/front/zsharpcurly/term.rs

```diff
@@ -0,0 +1,1363 @@
+    match &u.ty {
+        Ty::Integer => Ok(T::new(
+            Ty::Field,
+            term![Op::IntToPf(default_field()); u.term],
+        )),
+        u => Err(format!("Cannot do int-to-field on {u}")),
+    }
+}
+
+pub fn field_to_integer(u: T) -> Result<T, String> {
+    match &u.ty {
+        Ty::Field => Ok(T::new(Ty::Integer, term![Op::PfToInt; u.term])),
+        u => Err(format!("Cannot do int-to-field on {u}")),
+    }
+}
+
+pub fn int_to_bits(i: T, n: usize) -> Result<T, String> {
+    match &i.ty {
+        Ty::Integer => uint_to_bits(T::new(Ty::Uint(n), term![Op::IntToBv(n); i.term])),
+        u => Err(format!("Cannot do uint-to-bits on {u}")),
+    }
+}
+
+pub fn int_size(i: T) -> Result<T, String> {
+    match &i.ty {
+        Ty::Integer => Ok(T::new(Ty::Uint(32), term![Op::IntSize; i.term])),
+        u => Err(format!("Cannot do sizeof on {u}")),
+    }
+}
+
+pub fn int_modinv(i: T, m: T) -> Result<T, String> {
+    match (&i.ty, &m.ty) {
+        (Ty::Integer, Ty::Integer) => Ok(T::new(
+            Ty::Integer,
+            term![Op::IntBinOp(IntBinOp::ModInv); i.term, m.term],
+        )),
+        u => Err(format!("Cannot do modinv on {:?}", u)),
+    }
+}
+
+pub fn uint_to_uint(u: T, w: usize) -> Result<T, String> {
+    match &u.ty {
+        Ty::Uint(n) if *n <= w => Ok(T::new(Ty::Uint(w), term![Op::BvUext(w - n); u.term])),
+        Ty::Uint(n) => Err(format!("Tried narrowing uint{n}-to-uint{w} attempted")),
+        u => Err(format!("Cannot do uint-to-uint on {u}")),
+    }
+}
+
+pub fn uint_to_bits(u: T) -> Result<T, String> {
+    match &u.ty {
+        Ty::Uint(n) => Ok(T::new(
+            Ty::Array(*n, Box::new(Ty::Bool)),
+            ir_array(
+                Sort::Bool,
+                (0..*n).rev().map(|i| term![Op::BvBit(i); u.term.clone()]),
+            ),
+        )),
+        u => Err(format!("Cannot do uint-to-bits on {u}")),
+    }
+}
+
+// XXX(rsw) is it correct to enforce length here, vs. in (say) builtin_call in mod.rs?
+pub fn uint_from_bits(u: T) -> Result<T, String> {
+    match &u.ty {
+        Ty::Array(bits, elem_ty) if **elem_ty == Ty::Bool => match bits {
+            8 | 16 | 32 | 64 => Ok(T::new(
+                Ty::Uint(*bits),
+                term(
+                    Op::BvConcat,
+                    u.unwrap_array_ir()?
+                        .into_iter()
+                        .map(|z: Term| -> Term { term![Op::BoolToBv; z] })
+                        .collect(),
+                ),
+            )),
+            l => Err(format!("Cannot do uint-from-bits on len {l} array")),
+        },
+        u => Err(format!("Cannot do uint-from-bits on {u}")),
+    }
+}
+
+pub fn field_to_bits(f: T, n: usize) -> Result<T, String> {
+    match &f.ty {
+        Ty::Field => uint_to_bits(T::new(Ty::Uint(n), term![Op::PfToBv(n); f.term])),
+        u => Err(format!("Cannot do uint-to-bits on {u}")),
+    }
+}
+
+pub fn field_to_bool_unsafe(f: T) -> Result<T, String> {
```

Comment:
@alex-ozdemir, can you review this one?

Reply:
I think that you actually want [this operator](https://github.com/circify/circ/blob/master/src/ir/term/mod.rs#L201) using it will maintain the connection between the input and the output of the operator in the final constraint system, it will just assume that the value is 0 or 1. See [this part](https://github.com/circify/circ/blob/master/src/target/r1cs/trans.rs#L556) of R1CS lowering.

What do you think?

Reply:
Thanks, that makes sense! Should we still wrap it in a new_witness Op? If we do so, in the test, should we also pass the new_witness value?

Reply:
Fixed it